Fix automation crash

michaelgregorius opened the pull request:
Fix for the crash that occurs when the zooming or snapping model of the Song Editor is used for automation.

Resolves #6711

michaelgregorius committed:
Fix automation crash (#6711)

Fix a crash that occurs if the zooming or snapping model of the Song
Editor is used for automation.

The crash is caused by a failed static_cast to a Model* in
Model::parentModel(). The cast fails because in the constructor of
SongEditor the SongEditor is set as the parent of the zooming and
snapping model:
m_zoomingModel->setParent(this);
m_snappingModel->setParent(this);

This commit is rather a "band aid" fix because it only fixes the crash
by changing the static_cast to a dynamic_cast. However this means that
the name of the automation clip is initially empty.

A better solution would be to not use the Qt parent/child relationships
to implement the model hierarchies.

michaelgregorius committed:
Move implementations of Model into cpp file (#6711)

Move the implementation of the Model methods into Model.cpp so that
recompiles after changes are much quicker. Make Model::
isDefaultConstructed const.

superpaik commented:
Coding style: I think you should remove underscore on function parameters.
Other than that it works ok.

michaelgregorius committed:
Code review changes (#6711)

Remove underscores and whitespace in the Model files.

michaelgregorius committed:
Simplify logic of Model::fullDisplayName (#6711)

Simplify the logic of the method Model::fullDisplayName.

Please note that this shows that with the current logic if the parent
model has a non-empty name like "parentmodel" and the name of the child
model is empty the result is the name "parentmodel >" which might not be
what's wanted.

michaelgregorius commented:
Underscores and whitespace has been removed with commit fda938fc.

I have also simplified the logic of `Model::fullDisplayName` with commit 9fe7fbd6. This might also have made more obvious a potential problem if the parent has a non-empty name "xyz" but the child has an empty one. This will result in the name "xyz >" being returned.

messmerd commented on include/Model.h:
```diff
@@ -37,35 +37,18 @@ class LMMS_EXPORT Model : public QObject
-		m_defaultConstructed( _default_constructed )
-	{
-	}
+	Model(Model * parent, QString displayName = QString(),
+		  bool defaultConstructed = false );
```
```suggestion
	Model(Model* parent, QString displayName = QString(),
		  bool defaultConstructed = false);
```

messmerd commented on src/core/Model.cpp:
```diff
@@ -27,22 +27,47 @@
 namespace lmms
 {
 
+Model::Model(Model * parent, QString displayName, bool defaultConstructed) :
```
```suggestion
Model::Model(Model* parent, QString displayName, bool defaultConstructed) :
```

messmerd commented on src/core/Model.cpp:
```diff
@@ -27,22 +27,47 @@
+
+Model* Model::parentModel() const
+{
+	return dynamic_cast<Model *>(parent());
```
```suggestion
	return dynamic_cast<Model*>(parent());
```

messmerd commented on src/core/Model.cpp:
```diff
@@ -27,22 +27,47 @@
+	return m_displayName;
+}
+
+void Model::setDisplayName( const QString& displayName )
```
```suggestion
void Model::setDisplayName(const QString& displayName)
```

messmerd commented on src/core/Model.cpp:
```diff
@@ -27,22 +27,47 @@
-	if( parentModel() ) 
+	const QString n = displayName();
+
+	if(parentModel())
```
```suggestion
	if (parentModel())
```

messmerd reviewed (approved):
Just some very minor style suggestions.
Everything else looks good to me and it worked when I tested it. I've also confirmed that the logic you simplified is still correct.

michaelgregorius committed:
More whitespace adjustments (#6711)

More whitespace adjustments as proposed in the code review.

michaelgregorius commented:
Thanks for the code review and double check @messmerd! I have implemented all your proposals with commit 0ebc1894.
